# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
%conda install scipy

Retrieving notices: ...working... done
done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.7.4
  latest version: 23.11.0

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.11.0



## Package Plan ##

  environment location: /Users/minhnguyen/anaconda3

  added / updated specs:
    - scipy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libblas-3.9.0              |16_osx64_openblas          13 KB  conda-forge
    libcblas-3.9.0             |16_osx64_openblas          13 KB  conda-forge
    liblapack-3.9.0            |16_osx64_openblas          13 KB  conda-forge
    pooch-1.8.0                |     pyhd8ed1ab_0          51 KB  conda-forge
    scipy-1.10.1               |  py311h939689b_0        22.4 MB  conda-

In [2]:
import scipy.stats as st

In [3]:
# import numpy and pandas
import numpy as np 
import pandas as pd


# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [4]:
# Your code here:
salary = pd.read_csv('../data/Salaries.csv')

/var/folders/9z/v2g_zsyn4m11dzc06c73_bbm0000gn/T/ipykernel_2662/4152113303.py:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  salary = pd.read_csv('../data/Salaries.csv')


Examine the `salaries` dataset using the `head` function below.

In [5]:
# Your code here:
salary.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [6]:
# Your code here:
salary.isnull().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [7]:
# Your code here
salary.shape

(148654, 13)

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [8]:
# Your code here
salary.drop(columns = 'Notes', axis = 1, inplace=True)
salary.isnull().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [9]:
# Your code here
salary.Status.unique()

array([nan, 'PT', 'FT'], dtype=object)

array([nan, 'PT', 'FT'], dtype=object)

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [10]:
# Your code here
i = salary[salary['Status'].notna()].index.tolist()
salary = salary.loc[i, :]

In [11]:
# Your code here
salary.isnull().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [12]:
# Your code here
salary.info()

<class 'pandas.core.frame.DataFrame'>
Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


In [13]:
# Your code here
salary.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.0  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.0    82313.7  96570.66  339653.70         436224.36  2014   
110534         0.0   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.0   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.0  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.0    82313.7  96570.66  339653.70         436224.36  2014   
110534         0.0   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.0   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

Do any type conversions and reset the index.

In [14]:
# Your code here
salary['Id'] = salary['Id'].astype(object)
for i in ['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']:
    salary[i] = salary[i].astype(float)
salary.info()

<class 'pandas.core.frame.DataFrame'>
Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  object 
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  float64
 4   OvertimePay       38119 non-null  float64
 5   OtherPay          38119 non-null  float64
 6   Benefits          38119 non-null  float64
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(6), int64(1), object(5)
memory usage: 3.8+ MB


In [15]:
# Your code here
salary.reset_index(inplace=True)
salary.drop(columns = ['index'], inplace=True)
salary

Id          EmployeeName                     JobTitle    BasePay  \
0      110532           David Shinn               Deputy Chief 3  129150.01   
1      110533            Amy P Hart            Asst Med Examiner  318835.49   
2      110534  William J Coaker Jr.     Chief Investment Officer  257340.00   
3      110535        Gregory P Suhr              Chief of Police  307450.04   
4      110536  Joanne M Hayes-White       Chief, Fire Department  302068.00   
...       ...                   ...                          ...        ...   
38114  148646      Carolyn A Wilson    Human Services Technician       0.00   
38115  148648        Joann Anderson  Communications Dispatcher 2       0.00   
38116  148649           Leon Walker                    Custodian       0.00   
38117  148650         Roy I Tillery                    Custodian       0.00   
38118  148654             Joe Lopez   Counselor, Log Cabin Ranch       0.00   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0             0.00  342802.63  38780.04  471952.64         510732.68  2014   
1         10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2             0.00   82313.70  96570.66  339653.70         436224.36  2014   
3             0.00   19266.72  91302.46  326716.76         418019.22  2014   
4             0.00   24165.44  91201.66  326233.44         417435.10  2014   
...            ...        ...       ...        ...               ...   ...   
38114         0.00       0.00      0.00       0.00              0.00  2014   
38115         0.00       0.00      0.00       0.00              0.00  2014   
38116         0.00       0.00      0.00       0.00              0.00  2014   
38117         0.00       0.00      0.00       0.00              0.00  2014   
38118         0.00    -618.13      0.00    -618.13           -618.13  2014   

              Agency Status  
0      San Francisco     PT  
1      San Francisco     FT  
2      San Francisco     PT  
3      San Francisco     FT  
4      San Francisco     FT  
...              ...    ...  
38114  San Francisco     PT  
38115  San Francisco     PT  
38116  San Francisco     PT  
38117  San Francisco     PT  
38118  San Francisco     PT  

[38119 rows x 12 columns]

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [16]:
# Your code here
salary['TotalPayBenefits'] == salary['BasePay'] + salary['OvertimePay'] + salary['OtherPay'] + salary['Benefits']

0         True
1        False
2         True
3        False
4         True
         ...  
38114     True
38115     True
38116     True
38117     True
38118     True
Length: 38119, dtype: bool

0         True
1        False
2         True
3        False
4         True
         ...  
38114     True
38115     True
38116     True
38117     True
38118     True
Length: 38119, dtype: bool

What is the percetage of employees for which the previous assumption is not True?

In [17]:
# Your code here:
(salary['TotalPayBenefits'] == salary['BasePay'] + salary['OvertimePay'] + salary['OtherPay'] + salary['Benefits']).value_counts()/len(salary)*100

True     73.708649
False    26.291351
Name: count, dtype: float64

True     73.708649
False    26.291351
dtype: float64

There are different departments in the city. List all departments and the count of employees in each department.

In [18]:
# Your code here:
salary.groupby('JobTitle').agg({'Id':'count'}).rename(columns = {'Id':'Employees'})

Employees
JobTitle                                 
ACPO,JuvP, Juv Prob (SFERS)             1
ASR Senior Office Specialist           22
ASR-Office Assistant                   15
Account Clerk                          93
Accountant I                            2
...                                   ...
Wire Rope Cable Maint Sprv              1
Worker's Comp Supervisor 1              6
Worker's Compensation Adjuster         26
X-Ray Laboratory Aide                  35
Youth Comm Advisor                      1

[996 rows x 1 columns]

Employees
JobTitle                                 
ACPO,JuvP, Juv Prob (SFERS)             1
ASR Senior Office Specialist           22
ASR-Office Assistant                   15
Account Clerk                          93
Accountant I                            2
...                                   ...
Wire Rope Cable Maint Sprv              1
Worker's Comp Supervisor 1              6
Worker's Compensation Adjuster         26
X-Ray Laboratory Aide                  35
Youth Comm Advisor                      1

[996 rows x 1 columns]

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [19]:
# Your code here: (compute the "Hourly_Wage")
salary['Hourly_Wage'] = (salary['TotalPayBenefits']/50)/40
salary.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [21]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)
mean_hw_ft = salary[salary['Status'] == 'FT']['Hourly_Wage'].mean()
display(mean_hw_ft)
len(salary[salary['Status'] == 'FT'])

69.26420640727143

22334

69.264206407271

In [28]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees

#import scipy.stats as st
from scipy.stats import ttest_1samp

# manual calculation
t_statistic = (np.mean(salary[salary['Status']=='FT']['Hourly_Wage']) - 75)/(np.std(salary[salary['Status']=='FT']['Hourly_Wage'], ddof = 1)/np.sqrt(len(salary[salary['Status']=='FT']['Hourly_Wage'])))
t_statistic



-35.80631941460521

-35.80631941460788

In [29]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
lower_ct = st.t.ppf((0.05/2),df= len(salary[salary['Status']=='FT']['Hourly_Wage'])-1)
upper_ct = st.t.ppf((1-(0.05/2)),df= len(salary[salary['Status']=='FT']['Hourly_Wage'])-1)
print('lower_ct:', lower_ct)
print('upper_ct:', upper_ct)

lower_ct: -1.9600702128845597
upper_ct: 1.9600702128845593


In [30]:
# Method 2: Use the p-value method.
# Your code here:
p = st.t.cdf(t_statistic,df= len(salary[salary['Status']=='FT']['Hourly_Wage'])-1 )
print(p)
p*2<0.025

2.2589156169228082e-273


True

In [31]:
p*2

4.5178312338456165e-273

In [32]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
# use ttest_1samp
stat, pval = ttest_1samp(salary[salary['Status'] == 'FT']['Hourly_Wage'], 75,alternative = 'two-sided')
print(t_statistic)
print( stat, pval)

-35.80631941460521
-35.80631941460526 4.517831233839447e-273


Are all the methods in agreement?

### Answer: yes, they are all in agreement.
As t_statistic < tcritical (-35.8 < - 1.96), we reject the H0: which is the hourly_wage is not significantly different than 75dollar/hr

We are also curious about salaries in the police force. The chief of police in San Francisco claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [136]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees
salary_police = salary[salary['JobTitle'].apply(lambda x: 'y' if 'Police' in x else 'n')== 'y']

# The claim is that salaries this year are higher than last year's mean of $86000/year for all salaried employees
# null hypothesis H0: salary mean of police force less than 86000/year, confidence 95% -> significant a = 0.05, one-tailed t-test (left)
t = (np.mean(salary_police['TotalPayBenefits']) - 86000)/(np.std(salary_police['TotalPayBenefits'], ddof = 1)/np.sqrt(len(salary_police['TotalPayBenefits'])))
t

50.252984742109966

50.252984742109916

In [137]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
tc = st.t.ppf(0.05, (len(salary_police['TotalPayBenefits']-1)))
tc
# -> tc = -1.65 < t = 50.25: We accept the H0 



-1.6456555819760406

In [138]:
st.t.cdf(-1.6456560042647304, (len(salary_police['TotalPayBenefits'])-1))

0.05000000000000005

In [139]:
# Method 2: Use the p-value method.
# Your code here:
p_val = st.t.cdf(t, (len(salary_police['TotalPayBenefits'])-1))
p_val

# p_val = 1.0 > a = 0.05 : we accept the H0 -> meaning: salary mean of police force less than 86000/year this year, with a confidence level of 95%, at significant alpha 0.05

1.0

In [141]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
from scipy.stats import ttest_1samp
ttest_1samp(salary_police['TotalPayBenefits'], 86000, alternative = 'less')



TtestResult(statistic=50.252984742109945, pvalue=1.0, df=1900)

Ttest_1sampResult(statistic=50.252984742109945, pvalue=1.0)

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [87]:
# Your code here: (Get the department which has most employees)
dep = salary.groupby('JobTitle').agg({'Id':'count'}).rename(columns = {'Id':'Employees'}).idxmax()['Employees']

'Transit Operator'

'Transit Operator'

In [89]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees
# H0: hourly wage of employees from 'Transit Operator' department is more than 35 dollars/hour, confidence level 95%, alpha 0.05
t = (np.mean(salary[salary['JobTitle']==dep]['Hourly_Wage'])-35) / (np.std(salary[salary['JobTitle']==dep]['Hourly_Wage'])/np.sqrt(len(salary[salary['JobTitle']==dep]['Hourly_Wage'])-1))
t


19.06499032490636

19.06499032490655

In [90]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
tc = st.t.ppf((1-0.05), (len(salary[salary['JobTitle']==dep]['Hourly_Wage'])-1))
tc

# t = 19.06 > tc = 1.65: We reject H0

1.645468777247024

In [92]:
# Method 2: Use the p-value method.
# Your code here:
p_value = 1 - st.t.cdf(t, (len(salary[salary['JobTitle']==dep]['Hourly_Wage'])-1))

# p_value =0.0 < a = 0.05 : We reject H0

0.0

In [98]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

from scipy.stats import ttest_1samp
ttest_1samp(salary[salary['JobTitle']==dep]['Hourly_Wage'], 35, alternative = 'greater')

TtestResult(statistic=19.064990324906383, pvalue=5.012771694204206e-76, df=2478)

Ttest_1sampResult(statistic=19.064990324906383, pvalue=5.012771694204206e-76)

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [110]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:
confidence_level = 0.95
degrees_freedom = len(salary)-1
hourly_wage_mean = np.mean(salary['Hourly_Wage'])
hourly_wage_sem = st.sem(salary['Hourly_Wage'])

confidence_interval = st.t.interval(confidence_level,
                                             degrees_freedom,
                                             hourly_wage_mean,
                                             hourly_wage_sem)
display('Hourly wage mean: {:.2f}'.format(hourly_wage_mean))
display('Confidence_interval: [{:.2f},{:.2f}]'.format(confidence_interval[0],confidence_interval[1]))

'Hourly wage mean: 50.13'

'Confidence_interval: [49.80,50.46]'

Sample mean: 50.13
Confidence interval: [49.80,50.46]


Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [111]:
# Your code here:
confidence_level = 0.95
degrees_freedom = len(salary[salary['Police']=='y'])-1
police_hourly_wage_mean = np.mean(salary[salary['Police']=='y']['Hourly_Wage'])
police_hourly_wage_sem = st.sem(salary[salary['Police']=='y']['Hourly_Wage'])

confidence_interval = st.t.interval(confidence_level,
                                             degrees_freedom,
                                             police_hourly_wage_mean,
                                             police_hourly_wage_sem)
display('Hourly wage mean: {:.2f}'.format(police_hourly_wage_mean))
display('Confidence_interval: [{:.2f},{:.2f}]'.format(confidence_interval[0],confidence_interval[1]))

'Hourly wage mean: 74.17'

'Confidence_interval: [72.96,75.39]'

Sample mean: 74.17
Confidence interval: [72.96,75.39]


# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [123]:
# Your code here: (Create the function)
def group_JobTitle (df):
    df2 = df.copy()
    if 'JobTitle' in df2.columns:
        df2['employee_group'] = df2['JobTitle'].apply(lambda x: 'Policemen' if 'Police' in x \
                                                      else ('Firemen' if 'Fire' in x \
                                                            else ('Medical' if 'Med'in x or 'Nurse' in x \
                                                                  else ('Lawyer' if 'Attorney' in x \
                                                                    else 'Other'))))
    else:
        print('There is no JobTitle column to group')
    return df2


In [125]:
# Your code here: (Apply the function)
salary_1 = group_JobTitle(salary)
salary_1.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage Police employee_group  
0  San Francisco     PT   255.366340      n          Other  
1  San Francisco     FT   239.826105      n        Medical  
2  San Francisco     PT   218.112180      n          Other  
3  San Francisco     FT   209.009610      y      Policemen  
4  San Francisco     FT   208.717550      n        Firemen

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage employee_group  
0  San Francisco     PT   255.366340          Other  
1  San Francisco     FT   239.826105        Medical  
2  San Francisco     PT   218.112180          Other  
3  San Francisco     FT   209.009610      Policemen  
4  San Francisco     FT   208.717550        Firemen

Determine how many "PT" and "FT" employess have all the employees groups.

In [176]:
# Your code here: (Store the output dataframe into a new variable)
#salary_1.groupby(['employee_group', 'Status']).agg({'Id':'count'}).reset_index().pivot(index='employee_group', columns='Status', values = 'Id')
#salary_1.pivot_table(index = ['employee_group', 'Status'], values = ['Id'], aggfunc= {'Id': 'count'}).reset_index().pivot(index = 'employee_group', columns = 'Status', values='Id')
data = pd.crosstab(salary_1['employee_group'], salary_1['Status'])


Status             FT     PT
employee_group              
Firemen          1333    178
Lawyers           317    102
Medical          1028   2889
Other           18126  12245
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [ ]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:
frequencies = []


In [187]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"
firemen_ft = ((1333+178)/data.sum().sum())*(data['FT'].sum()/data.sum().sum())*data.sum().sum()
firemen_ft

885.2979878800596

885.2979878800596

In [200]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"
firemen_pt = ((1333+178)/data.sum().sum())*(data['PT'].sum()/data.sum().sum())*data.sum().sum()
firemen_pt

625.7020121199403

625.7020121199403

In [199]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"
lawyer_ft = ((data.loc['Lawyer'].sum())/data.sum().sum())*(data['FT'].sum()/data.sum().sum())*data.sum().sum()
lawyer_ft

245.49295626852748

245.49295626852748

In [201]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"
lawyer_pt = ((data.loc['Lawyer'].sum())/data.sum().sum())*(data['PT'].sum()/data.sum().sum())*data.sum().sum()
lawyer_pt

173.5070437314725

173.5070437314725

In [203]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"
medical_ft = ((data.loc['Medical'].sum())/data.sum().sum())*(data['FT'].sum()/data.sum().sum())*data.sum().sum()
medical_ft

2294.9783047823917

2294.9783047823917

In [204]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"
medical_pt = ((data.loc['Medical'].sum())/data.sum().sum())*(data['PT'].sum()/data.sum().sum())*data.sum().sum()
medical_pt

1622.021695217608

1622.021695217608

In [205]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"
other_ft = ((data.loc['Other'].sum())/data.sum().sum())*(data['FT'].sum()/data.sum().sum())*data.sum().sum()
other_ft

17794.430966184842

17794.430966184842

In [206]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"
other_pt = ((data.loc['Other'].sum())/data.sum().sum())*(data['PT'].sum()/data.sum().sum())*data.sum().sum()
other_pt

12576.569033815158

12576.569033815158

In [207]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"
policemen_ft = ((data.loc['Policemen'].sum())/data.sum().sum())*(data['FT'].sum()/data.sum().sum())*data.sum().sum()
policemen_ft

1113.7997848841785

1113.7997848841785

In [208]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"
policemen_pt = ((data.loc['Policemen'].sum())/data.sum().sum())*(data['PT'].sum()/data.sum().sum())*data.sum().sum()
policemen_pt

787.2002151158215

787.2002151158215

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [209]:
# Your code here:
freq = {'FT': [firemen_ft, policemen_ft, medical_ft, lawyer_ft, other_ft], 'PT': [firemen_pt, policemen_pt, medical_pt, lawyer_pt, other_pt] }
frequencies = pd.DataFrame(freq)
frequencies

FT            PT
0    885.297988    625.702012
1   1113.799785    787.200215
2   2294.978305   1622.021695
3    245.492956    173.507044
4  17794.430966  12576.569034

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [210]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.05
st.chi2_contingency(data)

Chi2ContingencyResult(statistic=2676.642333711905, pvalue=0.0, dof=4, expected_freq=array([[  885.29798788,   625.70201212],
       [  245.49295627,   173.50704373],
       [ 2294.97830478,  1622.02169522],
       [17794.43096618, 12576.56903382],
       [ 1113.79978488,   787.20021512]]))

(2676.642333711905, 0.0, 4, array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

In [216]:
# the expected frequencies do not agree with the ones obtained by the st.chi2_contingency()
# pvalue = 0.0 < a = 0.05 : reject H0 hypothesis which is the assumption that the observed frequencies in the contingency table are independence.
chi2_critical = st.chi2.ppf(0.95, 4)
chi2_critical

# chi2_critical = 9.49 < < chi2 = 2676.64: reject H0 --> there is a relationship between employee_group and status (FT/PT) 

9.487729036781154